src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_code.go

```diff
@@ -91,13 +93,14 @@ func (a *StoragePowerActorCode_I) WithdrawBalance(rt Runtime, amount actor.Token
 
 	h, st := a.State(rt)
 
-	currEntry, found := st.PowerTable()[minerID]
-	if !found {
-		rt.Abort("minerID not found.")
+	ret := st._ensurePledgeCollateralSatisfied(rt)
+	if !ret {
+		rt.Abort("insufficient pledge collateral.")
 	}
 
+	currEntry := st._safeGetPowerEntry(rt, minerID)
 	if currEntry.AvailableBalance() < amount {
-		rt.Abort("insufficient balance.")
+		rt.Abort("insufficient available balance.")
 	}
 
 	currEntry.Impl().AvailableBalance_ = currEntry.AvailableBalance() - amount
```

Comment:
Please satisfy the `// TODO: send funds to msgSender` now

Reply:
+1. These functionalities should now all be doable with the current runtime VM context API (`SendPropagatingErrors`, `ImmediateCaller`)